File: services/graphql-crm/src/models/notifications.rs
use std::sync::Arc;

use async_graphql::{ComplexObject, Context, SimpleObject, dataloader::Loader};
use chrono::{DateTime, Utc};
use graphql_auth::models::user;
use graphql_core::PostgresDataLoader;
use uuid::Uuid;

#[derive(Debug, Clone, Copy, PartialEq, Hash, Eq)]
pub struct PrimaryKey(pub Uuid);

#[derive(Clone, Debug, PartialEq, Eq, SimpleObject)]
#[graphql(complex)]
pub struct Model {
    pub id: Uuid,
    #[graphql(skip)]
    pub user_id: Uuid,
    pub message: String,
    pub is_read: Option<bool>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
    pub link: Option<String>,
}

#[ComplexObject]
impl Model {
    async fn user(&self, ctx: &Context<'_>) -> async_graphql::Result<user::Model> {
        todo!()
    }
}

impl Loader<PrimaryKey> for PostgresDataLoader {
    type Error = Arc<sqlx::Error>;
    type Value = Model;

    async fn load(
        &self,
        keys: &[PrimaryKey],
    ) -> Result<std::collections::HashMap<PrimaryKey, Self::Value>, Self::Error> {
        todo!()
    }
}

File: src/graphql/queries/crm/notifications.ts
import { graphql } from "@/lib/graphql/client";

export const getNotification = graphql(`
  query GetNotification($id: UUID!) {
    crm {
      notification(id: $id) {
        id
        message
        isRead
        link
        createdAt
        updatedAt
        user {
          id
          name
        }
      }
    }
  }
`);

export const getNotifications = graphql(`
  query GetNotifications($limit: Int!, $page: Int!) {
    crm {
      notifications(limit: $limit, page: $page) {
        id
        message
        isRead
        link
        createdAt
        updatedAt
        user {
          id
          name
        }
      }
    }
  }
`);
